storage/storage.c

```diff
@@ -472,79 +465,155 @@ static secbool is_not_wipe_code(const uint8_t *pin, size_t pin_len) {
-    ui_callback(ui_rem, progress, ui_message);
-  }
+  ui_progress(0);
 
   PBKDF2_HMAC_SHA256_CTX ctx = {0};
   pbkdf2_hmac_sha256_Init(&ctx, pin, pin_len, salt, salt_len, 1);
   for (int i = 1; i <= 5; i++) {
     pbkdf2_hmac_sha256_Update(&ctx, PIN_ITER_COUNT / 10);
-    if (ui_callback && ui_message) {
-      progress =
-          ((ui_total - ui_rem) * 1000 + i * DERIVE_SECS * 100) / ui_total;
-      ui_callback(ui_rem - i * DERIVE_SECS / 10, progress, ui_message);
-    }
+    ui_progress(i * PIN_DERIVE_SECS * 100);
   }
   pbkdf2_hmac_sha256_Final(&ctx, kek);
 
   pbkdf2_hmac_sha256_Init(&ctx, pin, pin_len, salt, salt_len, 2);
   for (int i = 6; i <= 10; i++) {
     pbkdf2_hmac_sha256_Update(&ctx, PIN_ITER_COUNT / 10);
-    if (ui_callback && ui_message) {
-      progress =
-          ((ui_total - ui_rem) * 1000 + i * DERIVE_SECS * 100) / ui_total;
-      ui_callback(ui_rem - i * DERIVE_SECS / 10, progress, ui_message);
-    }
+    ui_progress(i * PIN_DERIVE_SECS * 100);
   }
   pbkdf2_hmac_sha256_Final(&ctx, keiv);
 
-  ui_rem -= DERIVE_SECS;
+  ui_rem -= PIN_DERIVE_SECS;
   memzero(&ctx, sizeof(PBKDF2_HMAC_SHA256_CTX));
   memzero(&salt, sizeof(salt));
 }
+#endif
+
+#if USE_OPTIGA
+static void derive_kek_optiga(
+    const uint8_t optiga_secret[OPTIGA_PIN_SECRET_SIZE],
+    const uint8_t *ext_salt, uint8_t kek[SHA256_DIGEST_LENGTH],
+    uint8_t keiv[SHA256_DIGEST_LENGTH]) {
+  uint8_t salt[HARDWARE_SALT_SIZE + EXTERNAL_SALT_SIZE] = {0};
+  size_t salt_len = 0;
+
+  memcpy(salt + salt_len, hardware_salt, HARDWARE_SALT_SIZE);
+  salt_len += HARDWARE_SALT_SIZE;
+
+  if (ext_salt != NULL) {
+    memcpy(salt + salt_len, ext_salt, EXTERNAL_SALT_SIZE);
+    salt_len += EXTERNAL_SALT_SIZE;
+  }
+
+  PBKDF2_HMAC_SHA256_CTX ctx = {0};
+  pbkdf2_hmac_sha256_Init(&ctx, optiga_secret, OPTIGA_PIN_SECRET_SIZE, salt,
+                          salt_len, 1);
+  // TODO Do we even need this one iteration or can we use 0 or completely leave
+  // out pbkdf2_hmac_sha256_Update()?
+  pbkdf2_hmac_sha256_Update(&ctx, 1);
+  pbkdf2_hmac_sha256_Final(&ctx, kek);
+
+  pbkdf2_hmac_sha256_Init(&ctx, optiga_secret, OPTIGA_PIN_SECRET_SIZE, salt,
+                          salt_len, 2);
+  pbkdf2_hmac_sha256_Update(&ctx, 1);
+  pbkdf2_hmac_sha256_Final(&ctx, keiv);
+
+  memzero(&ctx, sizeof(ctx));
+  memzero(&salt, sizeof(salt));
+}
+#endif
+
+static void derive_kek_set(const uint8_t *pin, size_t pin_len,
+                           const uint8_t *storage_salt, const uint8_t *ext_salt,
+                           uint8_t kek[SHA256_DIGEST_LENGTH],
+                           uint8_t keiv[SHA256_DIGEST_LENGTH]) {
+#if USE_OPTIGA
+  uint8_t optiga_secret[OPTIGA_PIN_SECRET_SIZE] = {0};
+  // TODO consider completely skipping optiga_pin_set() and optiga_pin_verify()
+  // when pin_len == 0. It adds no security in this case. If the Optiga goes
+  // bust, then the user can at least use their Trezor without PIN to transfer
+  // their funds elsewhere for example.
```

Comment:
I would like to do this. Opinions @onvej-sl?

Reply:
This is an attack scenario in which the attacker benefits from the fact that the optiga is not used: The attacker has a flash dump but nothing else. In particular, the attacker has no physical access to the device.

What I don't know is how realistic this attack is.

Reply:
Interesting point, but the scenario seems unrealistic to me.